[Nathan Wong] iP

e0543978 opened the pull request:
<h3> DukeTown


Here are 3 reasons why DukeTown is a superior app: 
1. There are way cooler features in this app :zany_face:
1. It sounds like the word __*NukeTown*__
1. Refer to reasons 1 and 2  


Here are some of the cool features DukeTown owns:
* Will notify you every minute to remind you of your tasks
* Personalised responses
* Provides `help` feature for new users

How do you get your hands on DukeTown? Here's a checklist to help you with that.
- [x] Like the app
- [ ] [Download] the app
- [ ] Get Started!
- [ ] Give us a 5 star rating

Here are some unedited responses from __all__ our satisfied users so far:

Satisfied User 1:

> ~~Felt that personalised responses were rude. I got called lame for adding a homework task~~  
> I defintely enjoyed the personal touch to the personalised messages! Looking forward to my next minute notification  


```java
boolean enjoyDukeTown = true;
if (enjoyDukeTown) {
    System.out.println("Give us a 5 Star Rating!");
    }
```

Commit:
Add Gradle support

e0543978 committed:
Testing my First Commit. Removed default Duke greeting

e0543978 committed:
Level-1 increments

e0543978 committed:
Level-2 Increments

e0543978 committed:
Level-3 Increments. Created a Task class

e0543978 committed:
Created subclasses of Task Class. Added a static method addTask in Duke.java

e0543978 committed:
Level-4 Increments

e0543978 committed:
A-TextUiTesting Increments

e0543978 committed:
Level-5 Increments. Capable of handling minimal 2 errors

e0543978 committed:
Level-6 Increments

e0543978 committed:
Added listCommand, doneCommand and deleteCommand to Duke.java

e0543978 committed:
Change exception message

e0543978 committed:
Level-7 Increments.
Add save file function.

e0543978 committed:
Level-8 Increments test

e0543978 committed:
Add Level-8 Increments.

Add dates and times.

e0543978 committed:
Merge branch 'branch-Level-7'

e0543978 committed:
Merge branch 'branch-Level-8'

e0543978 committed:
Add A-MoreOOP Increments

e0543978 committed:
Add A-Packages Increments

e0543978 committed:
Add A-JUnit Increments

e0543978 committed:
Ignore Manifest file

e0543978 committed:
Add JavaDoc comments Increment

e0543978 committed:
Add Coding Standard Increment

e0543978 committed:
Add Level-9 Increments

e0543978 committed:
Merge branch 'branch-A-CodingStandard'

e0543978 committed:
Merge branch 'branch-A-JavaDoc'

e0543978 committed:
Merge branch 'branch-Level-9'

e0543978 commented:
<h3> DukeTown


Here are 3 reasons why DukeTown is a superior app: 
1. There are way cooler features in this app :zany_face:
1. It sounds like the word __*NukeTown*__
1. Refer to reasons 1 and 2  


Here are some of the cool features DukeTown owns:
* Will notify you every minute to remind you of your tasks
* Personalised responses
* Provides `help` feature for new users

How do you get your hands on DukeTown? Here's a checklist to help you with that.
- [x] Like the app
- [ ] [Download] the app
- [ ] Get Started!
- [ ] Give us a 5 star rating

Here are some unedited responses from __all__ our satisfied users so far:

Satisfied User 1:

> ~~Felt that personalised responses were rude. I got called lame for adding a homework task~~  
> I defintely enjoyed the personal touch to the personalised messages! Looking forward to my next minute notification  


```java
boolean enjoyDukeTown = true;
if (enjoyDukeTown) {
    System.out.println("Give us a 5 Star Rating!");
    }
```

e0543978 committed:
Group commands into a Command class

e0543978 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

rootkie commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,31 @@
+ * A subclass of the Task Object. 
+ * Includes a deadline date to complete task by.
+ * 
+ */
```
There is no need for empty line before and after javadoc. See example [here](https://se-education.org/guides/conventions/java/intermediate.html#comments)
```suggestion
/**
 * A subclass of the Task Object. 
 * Includes a deadline date to complete task by.
 */
```

rootkie commented on src/main/java/duke/DeadlineCommand.java:
```diff
@@ -0,0 +1,21 @@
+/**
+ *
+ * Command to add a Deadline Task to the existing list of tasks.
+ */
```
There is no need for empty line before and after javadoc. 
```suggestion
/**
 * Command to add a Deadline Task to the existing list of tasks.
 */
```

rootkie commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,31 @@
+    public Deadline(String description, boolean isDone, String by) {
+        super(description, isDone);
+        this.date = LocalDate.parse(by);
+    }
```
LocalDate.parse can throw DateTimeException. Probably should handle it here or put it in constructor definition.
```suggestion
    public Deadline(String description, String by) throws DateTimeException {
        super(description);
        this.date = LocalDate.parse(by);
    }

    public Deadline(String description, boolean isDone, String by) throws DateTimeException {
        super(description, isDone);
        this.date = LocalDate.parse(by);
    }
```

rootkie commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,126 @@
+    private String filePath;
+    public ArrayList<Task> taskArr;
+
+    public Storage(String filePath) {
+        this.filePath = filePath;
+        this.taskArr = new ArrayList<Task>();
+    }
```
Why do you have a filePath field if your writeToFile at line 120 writes to a hardcoded path? May be an oversight?
```java
    public void writeToFile(TaskList taskArr) throws IOException {
        FileWriter fw = new FileWriter("./data/duke.txt");
        for (int i = 0; i < taskArr.size(); i++) {
            fw.write(taskArr.get(i).toString() + System.lineSeparator());
        }
        fw.close();
    }
```

rootkie commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,88 @@
+public class Parser {
+    public Parser() {
+
+    }
```
Why is public empty constructor necessary here?

rootkie commented on src/main/java/duke/ListCommand.java:
```diff
@@ -0,0 +1,20 @@
+public class ListCommand extends Command{
+    ListCommand() {
+        
+    }
```
Why is this constructor needed? If I remember correctly this is the default constructor for java classes without specifying it.

rootkie commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,58 @@
+
+
+    public static void main(String[] args) {
+
+        Duke duke = new Duke("./data/duke.txt");
+        duke.run();
+
```
Should probably remove the additional empty lines to be inline with your other methods.
```suggestion
        Duke duke = new Duke("./data/duke.txt");
        duke.run();
```

rootkie commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,88 @@
+                } else {
+                    System.out.println(taskAdded);
+                    int spaceIndex = command.indexOf(" ");
+                    Task task = new Todo(command.substring(spaceIndex + 1));
```
Probably consider refactoring this line and following lines by adding explanatory variables to make your code more obvious.
```suggestion
                    String description = command.substring(spaceIndex + 1);
                    Task task = new Todo(description);
```

rootkie commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,88 @@
+                    throw new DukeException("Remember to enter deadline in this format:\"[deadline] [task] /by [date]\"");
+                } else {
+                    System.out.println(taskAdded);
+                    int spaceIndex = command.indexOf(" ");
+                    int slashIndex = command.indexOf("/by ");
+                    Task task = new Deadline(command.substring(spaceIndex + 1, slashIndex - 1), command.substring(slashIndex + 4));
```
You can refactor this part by using split
```suggestion
                    String[] args = command.split(" /by ");
                    String description = args[0];
                    String date = args[1];
                    Task task = new Deadline(description, date);
```

e0543978 committed:
Use CheckStyle to detect coding style violations

wanyu-l commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,35 @@
+
+    @Override
+    public String toString() {
+
+        return "D " + super.toString() + " | " + date.format(DateTimeFormatter.ofPattern("MMM d yyyy"));
+    }
```
I like how you made use of toString() method!
Maybe you have missed the empty line?

wanyu-l commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,42 @@
+
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+//import java.time.format.DateTimeParseException;
```
Perhaps it can be deleted instead of commented if it is not used by the code?

wanyu-l commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,42 @@
+    @Override
+    public String toString() {
+        return "E " + super.toString() + " | " + date.format(DateTimeFormatter.ofPattern("MMM d yyyy"));
+
```
Like in DeadLine.java, there is an empty line within the toString() method

wanyu-l commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,86 @@
+                String keyword = commandArr[1];
+                return new FindCommand(keyword);
+            } catch (ArrayIndexOutOfBoundsException e) {
+                System.out.println("Please insert a word after \"done\"");
```
The printed output may not help the user as the input is "find" and output is "...done"

wanyu-l reviewed (commented):
Looks good! 
I have gained some insights just by reading your code, only some minor stuff.

e0543978 committed:
Add Level-10 Increments

e0543978 committed:
Merge branch 'branch-Level-10'

e0543978 committed:
Fix Done, Delete and Find commands

e0543978 committed:
Fix Duke responses

e0543978 committed:
Change ByeCommand to SaveCommand

e0543978 committed:
Change ByeCommand to SaveCommand

e0543978 committed:
Improve code quality of Parser class and Parse method

Length of line when instantiating a new task in parse
method is too long.
Use of magic literals when parsing date as a string
might be difficult to read.

Let's,
* create local variables for description and date strings to
shorten the length of line
* replace magic literal with a named constant

e0543978 committed:
Include assertions in Parser and Storage classes

Assume that DATE_PARSE_INDEX is always 4
Assume that description will never be empty

Let's,
* Add assertions to define the existing assumptions
for these classes

e0543978 committed:
Merge pull request #1 from e0543978/branch-A-Assertions

Include assertions in Parser and Storage classes

e0543978 committed:
Merge branch 'master' of https://github.com/e0543978/ip into branch-A-Assertions

* 'master' of https://github.com/e0543978/ip:

e0543978 committed:
Merge branch 'master' of https://github.com/e0543978/ip

* 'master' of https://github.com/e0543978/ip:
  Include assertions in Parser and Storage classes

e0543978 committed:
Improve Code Quality

Happy path is not prominent

Let's,
* Make happy path more prominent
where possible

e0543978 committed:
Merge pull request #2 from e0543978/branch-A-CodeQuality

Improve Code Quality

e0543978 committed:
Merge branch 'master' of https://github.com/e0543978/ip

e0543978 committed:
Merge branch 'branch-A-Assertions'

e0543978 committed:
Add C-Statistics Increment

e0543978 committed:
Merge branch 'branch-BCD-Extension'

e0543978 committed:
Improve code readability by changing
long method to few short methods

e0543978 committed:
Add A-BetterGui increment

e0543978 committed:
Change colour of text background

e0543978 committed:
Attempt to fix jar

e0543978 committed:
Add more tests for different classes

e0543978 committed:
Add representative screenshot of product

e0543978 committed:
Add User Guide

e0543978 committed:
Set theme jekyll-theme-slate

e0543978 committed:
Add exit function

e0543978 committed:
Merge branch 'master' of https://github.com/e0543978/ip

e0543978 committed:
Change Ui.png filename

e0543978 committed:
Attempt to fix Ui.png filename

e0543978 committed:
Change Ui responses
